File: ios/ReflectModule.swift
import Foundation
import React
import ReflectCore   // the shared, wrapper-agnostic engine (standalone pod)

// ─────────────────────────────────────────────────────────────────────────────
//  Thin React Native bridge over the shared, platform-agnostic ReflectCore engine.
//
//  ALL SDK logic — sessions, durable queue, HMAC signing, batching, response-driven
//  retry, dedup, device collection, deferred deep links, attribution, SKAN, ATT —
//  lives in ReflectCore, the SAME engine the Flutter + Unity wrappers bind to. This
//  module ONLY translates between the RN bridge (a single generic `handle` +
//  Promise + RCTEventEmitter) and the core's command-dispatch (`handle`) + listener
//  surface. The old standalone (unsigned NSURLSession / own IDFA / SKAN /
//  install_uuid) implementation is deleted.
// ─────────────────────────────────────────────────────────────────────────────

@objc(ReflectModule)
public class ReflectModule: RCTEventEmitter, ReflectListener {

    private static let deepLinkEvent = "ReflectDeepLink"
    private static let attributionEvent = "ReflectAttribution"

    // Shared reference so the host AppDelegate hooks (handleURL) can reach the core.
    private static weak var shared: ReflectModule?

    private let core = ReflectCore()
    private var hasListeners = false
    // Buffer one of each until JS subscribes — RCTEventEmitter DROPS (and warns on)
    // events emitted before startObserving. The core also buffers the deferred deep
    // link / first attribution until setListener; this covers the RN-subscribe window.
    private var pendingDeepLink: String?
    private var pendingAttribution: String?

    public override init() {
        super.init()
        // The core holds the listener WEAK; the bridge retains this RCTEventEmitter
        // module so `self` outlives the core's reference. Keep setListener in init.
        core.setListener(self)
        ReflectModule.shared = self
    }

    @objc public override static func requiresMainQueueSetup() -> Bool { return false }

    @objc public override func supportedEvents() -> [String] {
        return [ReflectModule.deepLinkEvent, ReflectModule.attributionEvent]
    }

    public override func startObserving() {
        hasListeners = true
        if let d = pendingDeepLink {
            sendEvent(withName: ReflectModule.deepLinkEvent, body: ["data": d]); pendingDeepLink = nil
        }
        if let a = pendingAttribution {
            sendEvent(withName: ReflectModule.attributionEvent, body: ["data": a]); pendingAttribution = nil
        }
    }

    public override func stopObserving() { hasListeners = false }

    // ─────────────────────────────────────────────────────────────────────────
    //  The single generic dispatcher (exposed to JS as NativeModules.ReflectModule
    //  .handle). Fire-and-forget commands resolve nil; async ones (verifyPurchase,
    //  resolveDeepLink, getAttributionWithTimeout, deleteUserData, requestIosTracking,
    //  updateConversionValue) resolve on the core's main-thread reply.
    // ─────────────────────────────────────────────────────────────────────────
    @objc(handle:args:resolve:reject:)
    func handle(_ method: String, args: NSDictionary?,
                resolve: @escaping RCTPromiseResolveBlock,
                reject: @escaping RCTPromiseRejectBlock) {
        core.handle(method: method, args: args as? [String: Any]) { value in
            if value is ReflectNotImplemented {
                reject("unimplemented", method, nil)
            } else if let e = value as? ReflectError {
                reject(e.code, e.message, nil)
            } else {
                resolve(ReflectModule.toResolvable(value))
            }
        }
    }

    // RN promises carry JSON-ish values. The core's sync getters return String /
    // Bool / nil directly; the few that return a dictionary (getDebugState,
    // verifyPurchase) are JSON-stringified so JS can JSON.parse them — the same
    // string-in/parse-out convention the JS uses for getAttribution/getInitialDeepLink.
    private static func toResolvable(_ value: Any?) -> Any? {
        guard let value = value else { return nil }
        if let dict = value as? [String: Any],
           let data = try? JSONSerialization.data(withJSONObject: dict),
           let s = String(data: data, encoding: .utf8) {
            return s
        }
        return value
    }

    // ── ReflectListener → RN device events (core dispatches these on the main thread)
    public func onDeepLink(_ data: Any) {
        let json = ReflectModule.jsonString(data)
        if hasListeners { sendEvent(withName: ReflectModule.deepLinkEvent, body: ["data": json]) }
        else { pendingDeepLink = json }
    }

    public func onAttribution(_ data: Any) {
        let json = ReflectModule.jsonString(data)
        if hasListeners { sendEvent(withName: ReflectModule.attributionEvent, body: ["data": json]) }
        else { pendingAttribution = json }
    }

    private static func jsonString(_ data: Any) -> String {
        if let dict = data as? [String: Any],
           let d = try? JSONSerialization.data(withJSONObject: dict),
           let s = String(data: d, encoding: .utf8) {
            return s
        }
        return String(describing: data)
    }

    // ─────────────────────────────────────────────────────────────────────────
    //  Host AppDelegate deep-link hooks. RN has no addApplicationDelegate auto-hook,
    //  so the host forwards openURL / continueUserActivity into the shared core via
    //  ReflectModule.handleURL, and stashes the cold-launch URL (so getInitialDeepLink
    //  resolves before JS subscribes). See README for the AppDelegate snippet.
    // ─────────────────────────────────────────────────────────────────────────
    @objc public static func handleURL(_ url: URL) {
        shared?.core.handleIncomingURL(url)
    }

    @objc public static func stashLaunchURL(_ url: URL) {
        UserDefaults.standard.set(url.absoluteString, forKey: "reflect_launch_url")
    }
}

File: android/src/main/kotlin/com/reflect/rn/ReflectModule.kt
package com.reflect.rn

import android.app.Activity
import android.content.Intent
import com.facebook.react.bridge.ActivityEventListener
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.reflect.core.ReflectCore
import com.reflect.core.ReflectListener
import com.reflect.core.ReflectResult
import org.json.JSONObject

// ─────────────────────────────────────────────────────────────────────────────
//  Thin React Native bridge over the shared com.reflect.core.ReflectCore engine.
//
//  ALL SDK logic — sessions, durable queue, HMAC signing, batching, response-driven
//  retry, dedup, device collection, deferred deep links, attribution — lives in
//  ReflectCore, the SAME engine the Flutter and Unity wrappers bind to. This module
//  ONLY translates between the RN bridge (a single generic `handle` + Promise +
//  DeviceEventEmitter) and the core's command-dispatch (`handle`) + listener
//  surface. It carries NO transport/attribution logic of its own — the old
//  standalone (unsigned HTTP, own install_uuid/GAID) implementation is deleted.
//
//  Written in Kotlin (not Java) so it compiles in the SAME kotlinc pass as the
//  source-included core — a Java module can't see the same-module Kotlin core.
// ─────────────────────────────────────────────────────────────────────────────
class ReflectModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext),
    ReflectListener, ActivityEventListener, LifecycleEventListener {

    // APPLICATION context (never an Activity) — mirrors the Flutter plugin.
    private val core = ReflectCore(reactContext.applicationContext)

    // Buffer one of each until the JS side (NativeEventEmitter) is live — RN drops
    // events emitted with no active React instance. The core already buffers the
    // deferred deep link / first attribution until setListener; this second buffer
    // covers the separate window before the JS instance connects.
    private var pendingDeepLink: String? = null
    private var pendingAttribution: String? = null

    init {
        core.setListener(this)
        core.start()                                  // durable-queue reload
        reactContext.addActivityEventListener(this)
        reactContext.addLifecycleEventListener(this)
    }

    override fun getName(): String = "ReflectModule"

    // ── The single generic dispatcher — every host API call funnels through here
    //    and hits the core's 41-case handle(). Fire-and-forget commands resolve
    //    null; async ones resolve on the core's main-thread reply.
    @ReactMethod
    fun handle(method: String, args: ReadableMap?, promise: Promise) {
        val m: Map<String, Any?>? = args?.toHashMap()
        core.handle(method, m, object : ReflectResult {
            override fun success(value: Any?) { promise.resolve(toResolvable(value)) }
            override fun error(code: String, message: String?, details: Any?) { promise.reject(code, message) }
            override fun notImplemented() { promise.reject("unimplemented", method) }
        })
    }

    // RN's Promise.resolve carries null/Boolean/Number/String/Writable*. The core's
    // sync getters return String/Boolean/null directly; the few that return a Map
    // (getDebugState, verifyPurchase) are JSON-stringified so JS can JSON.parse them.
    private fun toResolvable(value: Any?): Any? =
        if (value is Map<*, *>) JSONObject(value).toString() else value

    // RCTEventEmitter contract stubs (so NativeEventEmitter doesn't warn on the
    // legacy bridge). DeviceEventEmitter needs no real subscribe bookkeeping.
    @ReactMethod fun addListener(eventName: String) { /* no-op */ }
    @ReactMethod fun removeListeners(count: Int) { /* no-op */ }

    // ── Core listener → RN device events (core dispatches these on the main thread).
    override fun onDeepLink(data: Any) {
        val json = jsonString(data)
        if (!emitJson(EVENT_DEEP_LINK, json)) pendingDeepLink = json
    }

    override fun onAttribution(data: Any) {
        val json = jsonString(data)
        if (!emitJson(EVENT_ATTRIBUTION, json)) pendingAttribution = json
    }

    /** @return true if delivered, false if there was no active JS instance. */
    private fun emitJson(name: String, json: String): Boolean {
        if (!reactContext.hasActiveReactInstance()) return false
        val map = Arguments.createMap().apply { putString("data", json) }
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(name, map)
        return true
    }

    private fun flushPending() {
        pendingDeepLink?.let { if (emitJson(EVENT_DEEP_LINK, it)) pendingDeepLink = null }
        pendingAttribution?.let { if (emitJson(EVENT_ATTRIBUTION, it)) pendingAttribution = null }
    }

    private fun jsonString(data: Any): String =
        if (data is Map<*, *>) JSONObject(data).toString() else data.toString()

    // ── Activity + deep-link host wiring. Standard ReactActivity apps forward
    //    onNewIntent → RN → ActivityEventListener automatically, so WARM deep links
    //    are captured with no host code. Cold links are read by the core off the
    //    activity intent via getInitialDeepLink once the activity is attached.
    // NOTE: params are NON-NULL to match RN's ActivityEventListener. Recent RN
    // (Kotlin-migrated ReactAndroid) declares these as non-null `Intent`/`Activity`,
    // so a nullable `Intent?`/`Activity?` override "overrides nothing" and the module
    // fails to compile. Older RN exposes them as Java platform types, which accept a
    // non-null override too — so this signature is correct on every supported RN.
    override fun onNewIntent(intent: Intent) {
        reactContext.currentActivity?.intent = intent               // so a later getInitialDeepLink sees it
        intent.data?.let { core.handleDeepLinkUri(it) }
    }

    override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) { /* unused */ }

    override fun onHostResume() {
        reactContext.currentActivity?.let { core.attachActivity(it) } // enables getInitialDeepLink
        flushPending()
    }

    override fun onHostPause() { /* no-op */ }

    override fun onHostDestroy() { core.detachActivity() }

    companion object {
        private const val EVENT_DEEP_LINK = "ReflectDeepLink"
        private const val EVENT_ATTRIBUTION = "ReflectAttribution"
    }
}
